packages/sw-cacheable-response-behavior/src/lib/behavior.js

```diff
@@ -0,0 +1,121 @@
+/*
+ Copyright 2016 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import assert from '../../../../lib/assert';
+
+/**
+ * @memberof module:sw-cacheable-response-behavior
+ *
+ * @example
+ * // Used as an automatically invoked as "behavior" by a RequestWrapper:
+ *
+ * const cacheName = 'runtime-cache';
+ * const requestWrapper = new goog.runtimeCaching.RequestWrapper({
+ *   cacheName,
+ *   behaviors: [
+ *     new goog.cacheableResponse.Behavior({
+ *       statuses: [0, 200, 404]
+ *     })
+ *   ]
+ * });
+ *
+ * // Set up a route to match any requests made against the example.com domain.
+ * // The responses will be cached if the response code is 0, 200, or 404, and
+ * // will not be cached otherwise.
+ * const route = new goog.routing.RegExpRoute({
+ *   match: ({url}) => url.domain === 'example.com',
+ *   handler: new goog.runtimeCaching.StaleWhileRevalidate({requestWrapper})
+ * });
+ *
+ * @example
+ * // Explicitly invoked usage independent of the goog.routing framework, via
+ * // the isResponseCacheable() method:
+ *
+ * // TODO: Write sample code.
+ */
+class Behavior {
+  /**
+   * Creates a new `Behavior` instance, which stores configuration and logic
+   * to determine whether a `Response` object is cacheable or not.
+   *
+   * If multiple criteria are present (e.g. both `statuses` and `headers`), then
+   * the `Response` needs to meet all of the criteria to be cacheable.
+   *
+   * @param {Object} input The input object to this function.
+   * @param {Array<Number>} [input.statuses] The status codes that are
+   *        checked when determining whether a `Response` is cacheable.
+   * @param {Object<String,String>} [input.headers] The header values that are
+   *        checked when determining whether a `Response` is cacheable.
+   */
+  constructor({statuses, headers} = {}) {
```

Comment:
I think this is the difference between defining a lifecycle callback and then providing a behavior module that implements that callback.

The `cacheWillUpdate` lifecycle callback should be defined as broadly as possible, and it's a good idea to pass in the original `Request` to it. I've modified `RequestWrapper` to take that into account.

With `goog.cacheableResponse.Behavior`, we have a wrapper to make implementing `cacheWillUpdate` for some common scenarios (header values, status values) easier for developers. Folks don't have to use `goog.cacheableResponse.Behavior` if that's too limiting, though—they can pass in any object that implements `cacheWillUpdate` and returns a `boolean`.

I think that gives the flexibility to folks who need it.

I'm also happy to tweak the `goog.cacheableResponse.Behavior` implementation if it makes sense, but I think we already have a way to provide full flexibility for folks who need it.